### PR TITLE
Cache bust

### PR DESCRIPTION
**Pull request by vxcamiloxv**
Template middleware return a jade compile function
minified or not with unique hash by template, the client side
get a template hash from a list where live availables templates,
aditional set a max-age expiration for better render performance
in client-side and is more friendly with development because not
need build templates manually.

Closes: #1204

### CODE REVIEW

**Comment by vxcamiloxv** [comment]
I think this [validation](https://github.com/pump-io/pump.io/pull/1350/files#diff-8dc0fb1119587e3ecd8cc6a9c9969d05R191) is not necessary, because if change template in the server but the client request come with hash from cache will returns 404

File: lib/template.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            var templatePath = path.join(viewsPath, file),
+                stat = fs.lstatSync(templatePath);
+
+
+            if (stat.isFile() && !blackListRegexp.test(file) && /\.jade$/.test(file)) {
+                var templateName = path.basename(file, ".jade"),
+                    templateContent = fs.readFileSync(templatePath, {encoding: "utf8"}),
+                    // Compile jade template for client side
+                    templateFn = jade.compileClient(templateContent, {
+                        externalRuntime: false,
```
Comment (strugee) [inline lib/template.js]: What is this option? I don't see it in [the docs](https://pugjs.org/api/reference.html); maybe you meant `inlineRuntimeFunctions`?

Comment (vxcamiloxv) [reply lib/template.js]: In jade 1.11.x both options`externalRuntime` or `inlineRuntimeFunctions` not exist, see: https://github.com/pugjs/pug/tree/master/packages/pug-runtime https://github.com/pugjs/pug/commit/1eeb9e6ea37a638601cc5559c37bf86ea4749b13

File: lib/template.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+"use strict";
+
+var fs = require("fs"),
+    path = require("path"),
+    crypto = require("crypto"),
+    _ = require("underscore"),
```
[the remark this answers is not in the document]
Comment (vxcamiloxv) [reply lib/template.js]: made

File: lib/template.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            "^user",
+            "^xss-error",
+            "^layout"
+        ],
+        blackListRegexp = new RegExp("(" + blackList .join("|") + ")\.jade$"),
+        templateLicense = "// This file is auto generated and will be overwritten \n" +
+        "// @licstart  The following is the entire license notice for the \n" +
+        "//  JavaScript code in this page.\n" +
+        "// \n" +
+        "// Copyright 2011-2013, E14N https://e14n.com/\n" +
```
[the remark this answers is not in the document]
Comment (vxcamiloxv) [reply lib/template.js]: made

File: lib/template.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
+// template.js
+//
+// Template middleware with cache-buster
+//
+// Copyright 2012, E14N https://e14n.com/
```
[the remark this answers is not in the document]
Comment (vxcamiloxv) [reply lib/template.js]: made

File: lib/template.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n" +
+        "// See the License for the specific language governing permissions and\n" +
+        "// limitations under the License.\n" +
+        "// \n" +
+        "// @licend  The above is the entire license notice\n" +
+        "// for the JavaScript code in this page.\n\n" +
+        "// XXX: this needs to be broken up into 3-4 smaller modules \n";
+
+
+    // Get al files from views directory but only take all .jade files
```
[the remark this answers is not in the document]
Comment (vxcamiloxv) [reply lib/template.js]: made

**Comment by vxcamiloxv** [comment]
Memory is always more fast than read from disk (example swap vs zswap), reduce IO-Operations is always better for performance, this option is also more friendly for frontend developers and don't increment much the startup time only increment seconds more

https://www.backblaze.com/blog/whats-diff-ram-vs-storage/
https://stackoverflow.com/questions/25467993/node-js-performance-reading-files-once-or-on-every-request

**Comment by vxcamiloxv** [comment]
Memory is always more fast than read from disk (example swap vs zswap), reduce IO-Operations is always better for performance, this option is also more friendly for frontend developers and don't increment much the startup time only increment seconds more

https://www.backblaze.com/blog/whats-diff-ram-vs-storage/
https://stackoverflow.com/questions/25467993/node-js-performance-reading-files-once-or-on-every-request

**Comment by vxcamiloxv** [comment]
@strugee  I was thinking about other ways to make better this implementation and if you still thinking keep in memory could be have performance issues I have two options.

### Server side options

A) The `template` (should be rename to `cachebuster`) middleware currently read a template from disk and serve this to client in debug mode, I could be will keep this behavior in production but with pre-compile and minified templates (and other statics)

B) I'll modified the request path and remove the hash, Will delegate the response to `express-statics` middleware.


### Client side options.

A) Reads all statics files and make hash, keeps in memory (only name file and hash)
- For template, a `templates.js` with files names and hash for every file  or append to `Pump._templates` in `layout.jade` as JSON
- For other statics reads all from disk and generate hash, keeps in memory.

B) A build task (a `script` in package.json) for generate a `JS` file and requires as node dependency, the file could be have this structure:

```
exports {
    js: [
        {"file-name": "hash"}
    ],
    css: [
        {"file-name": "hash"}
    ],
    template: [
        {"file-name": "hash"}
    ]
}
```

- For templates same as option A
- For other statics generate the file url, like this:

```js
script(src=getFileHash("/javascript/dist/pump.min.js"))
```
and the results

```html
<script src="/javascript/dist/pump.min.hdg6252.js"></script>
````

### Generation options

A) Generate and minified all statics in for example `javascript/dist/` `and templates/dist/` by a `script` task in `package.json`

B) Generate, minified and saves all statics by server on startup (developer friendly)


What do you think?

**Comment by vxcamiloxv** [comment]
I close this PR because has 9 months, and I was thinking in a better and more generic approach for that (not only templates), I Will work over options the previous comment.

**Pull request closed without merging**